File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
n = int(input())  # 계단 개수
s = [int(input()) for _ in range(n)]

if n <= 2:  # 계단이 2개 이하일 때
    print(sum(s))
else:
    dp = [0] * n
    dp[0], dp[1] = s[0], s[0] + s[1] 
    for i in range(2, n): 
        dp[i] = max(dp[i - 3] + s[i - 1] + s[i], dp[i - 2] + s[i])
    print(dp[-1])

File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N;
    static int[] stair;
    static int[] dp;

    public static void main(String[] args) throws IOException{
        input();
        solve();
        System.out.print(dp[N]);
    }

    public static void solve(){
        dp[1] = stair[1];
        dp[2] = stair[1] + stair[2];
        dp[3] = Math.max(
            stair[1] + stair[3],
            stair[2] + stair[3]
        );

        for(int i=4; i<N+1; i++){
            dp[i] = Math.max(
                dp[i-2] + stair[i],
                dp[i-3] + stair[i-1] + stair[i]
            );
        }
  
    }

    public static void input() throws IOException{
        N = Integer.parseInt(br.readLine().trim());
        stair = new int[301];
        dp = new int[301];

        for(int i=1; i<N+1; i++){
            stair[i] = Integer.parseInt(br.readLine());
        }
    }
}